### PR TITLE
FEATURE: TMA-144 Using a new efficient synchronization of process schedules

### PR DESCRIPTION
**Pull request by vhtien**

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/347/ : SUCCESS in 32s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/282/ : FAILURE in 58s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/13/ : FAILURE in 2m 01s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/348/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/283/ : SUCCESS in 43m 09s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/14/ : FAILURE in 2m 37s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/381/ : FAILURE in 9s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/316/ : FAILURE in 2m 01s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/47/ : FAILURE in 7m 59s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/384/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/319/ : FAILURE in 2m 04s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/50/ : FAILURE in 8m 15s

File: lib/gooddata/lcm/actions/synchronize_etls.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+          results = []
+
+          client = params.gdc_gd_client
+          development_client = params.development_client
+
+          domain_name = params.organization || params.domain
+          domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
+
+          # we will use client_id to detect the context of etl synchronization. If we don't have client_id, it means we are in Release brick
+          has_client_ids = params
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_etls.rb]: Can we change the condition from `has_client_ids` to something like `target_domain_different`? Would it be more accurate?

File: lib/gooddata/lcm/actions/synchronize_etls.rb [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                            .synchronize
+                            .group_by do |info|
+                              info[:to].first[:client_id]
+                            end
+                            .keys
+                            .compact
+          if !has_client_ids.empty?
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_etls.rb]: Can we move both the condition branches into separate methods for better readability? Something like `#synchronize` and `#synchronize_using_sdk`.

File: lib/gooddata/lcm/actions/synchronize_etls.rb [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+                params.gdc_logger.info "Transferring Schedules, from project: '#{from.title}', PID: '#{from.pid}', to project: '#{to_project.title}', PID: '#{to_project.pid}'"
+                res = GoodData::Project.transfer_schedules(from, to_project).sort_by do |item|
+                  item[:status]
+                end
+
+                results += res.map do |item|
+                  schedule = item[:schedule]
+
+                  # TODO: Review this and remove if not required or duplicate (GOODOT_CUSTOM_PROJECT_ID vs CLIENT_ID)
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_etls.rb]: todo

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/385/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/320/ : FAILURE in 2m 17s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/51/ : FAILURE in 8m 06s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/395/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/333/ : SUCCESS in 42m 08s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/64/ : FAILURE in 12m 54s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/396/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/334/ : SUCCESS in 41m 54s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/65/ : FAILURE in 8m 43s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/402/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/340/ : FAILURE in 42m 33s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/71/ : FAILURE in 10m 26s

**Comment by vhtien** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/403/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/341/ : SUCCESS in 41m 55s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/72/ : FAILURE in 10m 00s

**Comment by vhtien** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/405/ : FAILURE in 11s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/343/ : SUCCESS in 41m 58s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/74/ : FAILURE in 10m 51s

File: lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+              synchronize.flat_map do |info|
+                info[:to].flat_map do |to|
+                  to[:pid]
+                end
+              end
+            )
+
+            res = GoodData::Helpers.symbolize_keys(res)
+            res[:syncedResult][:clients].flat_map do |item|
+              item = item[:client]
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb]: Why the assignment?

File: lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                segment: segment_id,
+                master_project: segment.master_project_id,
+                client_id: item[:id],
+                client_project: item[:project].split('/').last,
+                status: 'ok'
+              }
+            end
+          end
+
+          params.synchronize.each do |info|
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb]: Would you mind moving this to a separate method with an explanatory name? Say `update_schedule_parameters`?

File: lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb (hunk 4 above)
Comment (vhtien) [inline lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb]: I don't want to repeat access `:client`. See line 56 and 57

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/413/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/351/ : FAILURE in 43m 58s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/82/ : FAILURE in 12m 50s

**Comment by vhtien** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/414/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/352/ : FAILURE in 42m 05s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/83/ : FAILURE in 13m 29s

**Pull request merged**